org.lflang/src/org/lflang/generator/MainContext.java

```diff
@@ -0,0 +1,127 @@
+ * for every build process.
+ */
+public class MainContext implements LFGeneratorContext {
+    /**
+     * The indicator that shows whether this build
+     * process is canceled.
+     */
+    private final CancelIndicator cancelIndicator;
+    /** The {@code ReportProgress} function of {@code this}. */
+    private final ReportProgress reportProgress;
+    /** Whether the requested build is required to be complete. */
+    private final Mode mode;
+    /** The result of the code generation process. */
+    private GeneratorResult result = null;
+    private final Properties args;
+    private final boolean hierarchicalBin;  // FIXME: The interface would be simpler if this were part of {@code args}, and in addition, a potentially useful feature would be exposed to the user.
+    private final Function<FileConfig, ErrorReporter> constructErrorReporter;
+    private boolean hasConstructedErrorReporter;
+
+    /**
+     * Initialize the context of a build process whose cancellation is
+     * indicated by {@code cancelIndicator}
+     * @param mode The mode of this build process.
+     * @param cancelIndicator The cancel indicator of the code generation
+     *                        process to which this corresponds.
+     */
+    public MainContext(Mode mode, CancelIndicator cancelIndicator) {
+        this(
+            mode, cancelIndicator, (message, completion) -> {}, new Properties(), false,
+            fileConfig -> new DefaultErrorReporter()
+        );
+    }
+
+    /**
+     * Initialize the context of a build process whose cancellation is
+     * indicated by {@code cancelIndicator}
+     * @param mode The mode of this build process.
+     * @param cancelIndicator The cancel indicator of the code generation
+     *                        process to which this corresponds.
+     * @param reportProgress The {@code ReportProgress} function of
+     *                       {@code this}.
+     * @param args Any arguments that may be used to affect the product of the
+     *             build.
+     * @param hierarchicalBin Whether the bin directory should be structured
+     *                        hierarchically.
+     * @param constructErrorReporter A function that constructs the appropriate
+     *                               error reporter for the given FileConfig.
+     */
+    public MainContext(
+        Mode mode,
+        CancelIndicator cancelIndicator,
+        ReportProgress reportProgress,
+        Properties args,
+        boolean hierarchicalBin,
+        Function<FileConfig, ErrorReporter> constructErrorReporter
+    ) {
+        this.mode = mode;
+        this.cancelIndicator = cancelIndicator;
+        this.reportProgress = reportProgress;
+        this.args = args;
+        this.hierarchicalBin = hierarchicalBin;
+        this.constructErrorReporter = constructErrorReporter;
+        this.hasConstructedErrorReporter = false;
+    }
+
+    @Override
+    public CancelIndicator getCancelIndicator() {
```

Comment:
JavaDoc missing in the remainder of this file.

Reply:
As indicated by the `@Override` annotation, the JavaDoc here would just be a duplicate of the documentation that appears for the interface. The interface fully specifies the API.

It's no problem to add it -- it's just copy-paste -- but I just want to double check for confirmation that that is really what we want. It would leave us with more documentation to maintain, and the two copies would have to be kept consistent with each other. (For what it's worth, the Google style guide [allows JavaDoc to be omitted](https://google.github.io/styleguide/javaguide.html#s7.3.4-javadoc-non-required) when `@Override` is present.)